docs: add docs for setting up federation-frontend with Helm (#9735)
grafana/mimir: docs/sources/helm-charts/mimir-distributed/configure/configure-gem-federation-frontend.md

---

> title: "Configure GEM Federation-Frontend"

Author: I copied it from the GEM docs, but in reality we refer to all composite-name service with lowercase (e.g. query-frontend, store-gateway, etc.). I changed to "federation-frontend" in 025e8c2d9e9ce27277ebb73b8a5d1edfff56e091

menuTitle: "GEM Federation-Frontend"
description: "Learn how to enable the GEM Federation-Frontend."
---

# Configure GEM Federation-Frontend

This article explains how to deploy the Grafana Enterprise Metrics (GEM) Federation-Frontend in a Kubernetes cluster using Helm. The Federation-Frontend allows you to query data from multiple GEM clusters using a single endpoint. For more information about cluster query federation, refer to the [Federation-Frontend documentation](https://grafana.com/docs/enterprise-metrics/<GEM_VERSION>/operations/cluster-query-federation).

> This guide focuses specifically on deploying the Federation-Frontend component as a standalone deployment without any additional GEM or Mimir components.

Author: good call. Changed in 0e945e7eb5486043038a66c24c0551740aaf14db

## Deploy the GEM Federation-Frontend

1. Create a Kubernetes `Secret` named `gem-tokens` with the GEM access tokens for each of the remote GEM clusters. This `Secret` will be later used in the Helm values file. Replace `TOKEN1` and `TOKEN2` with the access tokens for the remote GEM clusters.

   ```yaml
   apiVersion: v1
   kind: Secret
   metadata:
     name: gem-tokens
   data:
     CLUSTER_1_GEM_TOKEN: TOKEN1
     CLUSTER_2_GEM_TOKEN: TOKEN2
   ```

2. Apply the secret to your cluster in the `federation-frontend-demo` namespace with the `kubectl` command:

   ```bash
   kubectl -n federation-frontend-demo apply -f mysecret.yaml
   ```

3. Create a values file named `federation-frontend.yaml` with the following content.

   Replace `http://gem-query-frontend.monitoring.svc.cluster.local:8080/prometheus` and `https://gem.monitoring.acme.local/prometheus` with the URLs of the GEM clusters you want to query. Replace `tenant-1` and `tenant-2` with the tenant IDs of the remote GEM clusters.

>    Note that the resource settings shown below are examples that should be sufficient for small deployments. Adjust the values based on your specific requirements and load:

Reviewer: ```suggestion
   Note that these resource settings are examples that are sufficient for small deployments. Adjust the values based on your specific requirements and load:
```

       # For a complete list of configuration options, refer to the configuration reference at https://grafana.com/docs/enterprise-metrics/<GEM_VERSION>/config/reference/#federation.
       multitenancy_enabled: false
       federation:
         proxy_targets:
           - name: "cluster-1"
             url: "http://gem-query-frontend.monitoring.svc.cluster.local:8080/prometheus"
             basic_auth:
               username: tenant-1
               password: "${CLUSTER_1_GEM_TOKEN}"

           - name: "cluster-2"
             url: "https://gem.monitoring.acme.local/prometheus"
             basic_auth:
               username: tenant-2
               password: "${CLUSTER_2_GEM_TOKEN}"

   # Disable MinIO
   minio:
     enabled: false

   # The federation-frontend doesn't need the rollout-operator for rollouts, so it can be disabled.
   rollout_operator:
     enabled: false
   ```

4. Deploy the Federation-Frontend using `helm`:

   ```bash
   helm install federation-frontend grafana/mimir-distributed -f federation-frontend.yaml -n federation-frontend-demo
   ```

>    This will deploy only the Federation-Frontend component. The Federation-Frontend will be configured to proxy queries to the GEM clusters specified in the `proxy_targets` configuration.

Reviewer: ```suggestion
   This only deploys the Federation-Frontend component. The Federation-Frontend is configured to proxy queries to the GEM clusters specified in the `proxy_targets` configuration.
```

5. Verify that the Federation-Frontend is running. The simplest way to do this is to issue a label names query against the Federation-Frontend service.

>    This example tries to reach the Kubernetes service from within the cluster.

Reviewer: ```suggestion
   This example tries to reach the Kubernetes service from the cluster.
```

---
title: "Configure GEM federation-frontend"
menuTitle: "GEM federation-frontend"
description: "Learn how to enable the GEM federation-frontend."
---

# Configure GEM federation-frontend

You can deploy the Grafana Enterprise Metrics (GEM) federation-frontend in a Kubernetes cluster using Helm. The federation-frontend allows you to query data from multiple GEM clusters using a single endpoint. For more information about cluster query federation, refer to the [federation-frontend documentation](https://grafana.com/docs/enterprise-metrics/<GEM_VERSION>/operations/cluster-query-federation).

{{< admonition type="note" >}}
This guide focuses specifically on deploying the federation-frontend component as a standalone deployment without any additional GEM or Mimir components.
{{< /admonition >}}

## Before you begin

> 1. Set up a GEM cluster: For information about setting up and configuring a GEM deployment, refer to the [Configure Grafana Enterprise Metrics]({{< relref "./configure-grafana-enterprise-metrics" >}}) and [Get started with Grafana Mimir using the Helm chart]({{< relref "../get-started-helm-charts" >}}) articles.

Reviewer: ```suggestion
1. Set up a GEM cluster: For information about setting up and configuring a GEM deployment, refer to [Configure Grafana Enterprise Metrics]({{< relref "./configure-grafana-enterprise-metrics" >}}) and [Get started with Grafana Mimir using the Helm chart]({{< relref "../get-started-helm-charts" >}}).
```

Reviewer: We switched to using full URLs and not relrefs for links, per https://grafana.com/docs/writers-toolkit/write/links/#link-to-grafanacom-pages.

Author: the cascade variable for the mimir-distributed docs isn't present and the rest of the helm docs use relrefs. I'll stick with relrefs for now

   ```yaml
   apiVersion: v1
   kind: Secret
   metadata:
     name: gem-tokens
   data:
     CLUSTER_1_GEM_TOKEN: TOKEN1
     CLUSTER_2_GEM_TOKEN: TOKEN2
   ```

2. Apply the secret to your cluster in the `federation-frontend-demo` namespace with the `kubectl` command:

   ```bash
   kubectl -n federation-frontend-demo apply -f mysecret.yaml
   ```

3. Create a values file named `federation-frontend.yaml` with the following content.

   Replace `http://gem-query-frontend.monitoring.svc.cluster.local:8080/prometheus` and `https://gem.monitoring.acme.local/prometheus` with the URLs of the GEM clusters you want to query. Replace `tenant-1` and `tenant-2` with the tenant IDs of the remote GEM clusters.

   Note that these resource settings are examples that are sufficient for small deployments. Adjust the values based on your specific requirements and load:

   ```yaml
   # Enable enterprise features
   enterprise:
     enabled: true

   # Enable and configure federation-frontend
   federation_frontend:
     enabled: true

>      # Since this is a standalone deployment, we configure the chart to not render any of the other GEM components.

Reviewer: ```suggestion
     # Since this is a standalone deployment, configure the chart to not render any of the other GEM components.
```

     disableOtherComponents: true
     replicas: 2
     resources:
       requests:
         cpu: 100m
         memory: 128Mi
       limits:
         cpu: 1
         memory: 256Mi
     extraEnvFrom:
       - secretRef:
           name: gem-tokens

   # Configure the remote GEM clusters to query.
   mimir:
     structuredConfig:
       auth:
         type: trust

>        # The federation-frontend doesn't handle authentication and authorization. Disabling multitenancy means the federation-frontend will not require the X-Scope-OrgID header.

Reviewer: ```suggestion
       # The federation-frontend doesn't handle authentication or authorization. Disabling multitenancy means the federation-frontend doesn't require the X-Scope-OrgID header.
```

>        # With disabled multitenant the federation-frontend will use the authn/z material from the proxy_targets configuration.

Reviewer: ```suggestion
       # With disabled multitenancy, the federation-frontend uses the authn/z material from the proxy_targets configuration.
```

       # For a complete list of configuration options, refer to the configuration reference at https://grafana.com/docs/enterprise-metrics/<GEM_VERSION>/config/reference/#federation.
       multitenancy_enabled: false
       federation:
         proxy_targets:
           - name: "cluster-1"
             url: "http://gem-query-frontend.monitoring.svc.cluster.local:8080/prometheus"
             basic_auth:
               username: tenant-1
               password: "${CLUSTER_1_GEM_TOKEN}"

           - name: "cluster-2"
             url: "https://gem.monitoring.acme.local/prometheus"
             basic_auth:
               username: tenant-2
               password: "${CLUSTER_2_GEM_TOKEN}"

   # Disable MinIO
   minio:
     enabled: false

   # The federation-frontend doesn't need the rollout-operator for rollouts, so it can be disabled.
   rollout_operator:
     enabled: false
   ```

4. Deploy the federation-frontend using `helm`:

   ```bash
   helm install federation-frontend grafana/mimir-distributed -f federation-frontend.yaml -n federation-frontend-demo
   ```

>    This will deploy only the federation-frontend component. The federation-frontend will be configured to proxy queries to the GEM clusters specified in the `proxy_targets` configuration.

Reviewer: ```suggestion
   This deploys only the federation-frontend component. The federation-frontend is configured to proxy queries to the GEM clusters that you specify in the `proxy_targets` configuration.
```

5. Verify that the federation-frontend is running. The simplest way to do this is to issue a label names query against the federation-frontend service.

>    This example tries to reach the Kubernetes service from the cluster and request the label names from the past 1 year.

Reviewer: ```suggestion
   This example tries to reach the Kubernetes service from the cluster and requests the label names from the past year.
```

   ```bash
   curl -XPOST 'https://mimir-federation-frontend:8080/prometheus/api/v1/labels' \
     -d "start=$(date -u +%Y-%m-%dT%H:%M:%S.0Z -d '1 year ago' 2>/dev/null || date -u -v -1y +%Y-%m-%dT%H:%M:%S.0Z)" \
     -d "end=$(date -u +%Y-%m-%dT%H:%M:%S.0Z)"

>    ```

Reviewer 2: One year seems like a lot of data to query - how about the last 12 hours by default?

   federation_frontend:
     enabled: true
     replicas: 2
     resources:
       requests:
         cpu: 100m
         memory: 128Mi
       limits:
         cpu: 1
         memory: 256Mi
     extraEnvFrom:
       - secretRef:
           name: gem-tokens

   # Configure the remote GEM clusters to query.
   mimir:
     structuredConfig:
       auth:
         type: trust
       # The federation-frontend doesn't handle authentication and authorization. Disabling multitenancy means the federation-frontend will not require the X-Scope-OrgID header.
       # With disabled multitenant the federation-frontend will use the authn/z material from the proxy_targets configuration.
       # For a complete list of configuration options, refer to the configuration reference at https://grafana.com/docs/enterprise-metrics/<GEM_VERSION>/config/reference/#federation.
       multitenancy_enabled: false
       federation:
         proxy_targets:
           - name: "cluster-local-gem"
             url: "http://gem-query-frontend.monitoring.svc.cluster.local:8080/prometheus"
             basic_auth:
               username: tenant-1
               password: "${CLUSTER_LOCAL_GEM_PASSWORD}"

>            - name: "cluster-remote-gem"

Reviewer 2: Perhaps cluster 1 and cluster 2 would be better names to use here? Might be confusing to refer to one as local and one as remote given there's no requirement about where the clusters are.

Author: 👍 done in https://github.com/grafana/mimir/pull/9735/commits/909a59625b30a74edfea842ce06e062d1e22347a

     enabled: false
   store_gateway:
     enabled: false
   ruler:
     enabled: false
   overrides_exporter:
     enabled: false

   # Disable all caches
   admin-cache:
     enabled: false
   chunks-cache:
     enabled: false
   index-cache:
     enabled: false
   metadata-cache:
     enabled: false
   results-cache:
     enabled: false

   # Disable MinIO
   minio:
     enabled: false

   # The federation-frontend doesn't need the rollout-operator for rollouts, so it can be disabled.
   rollout_operator:
     enabled: false

   tokengenJob:
     enable: false
   ```

4. Deploy the Federation-Frontend using `helm`:

   ```bash
   helm install federation-frontend grafana/mimir-distributed -f federation-frontend.yaml -n federation-frontend-demo
   ```

   This will deploy only the Federation-Frontend component. The Federation-Frontend will be configured to proxy queries to the GEM clusters specified in the `proxy_targets` configuration.

5. Verify that the Federation-Frontend is running. The simplest way to do this is to issue a label names query against the Federation-Frontend service.

   This example tries to reach the Kubernetes service from within the cluster.
   ```bash

>    curl -XPOST 'https://mimir-federation-frontend:8080/prometheus/api/v1/labels' -d 'start=2024-01-01T00:00:00.0Z' -d 'end=2025-01-01T00:00:00.0Z'

Reviewer 2: This will query all label names for the year 2024 - is there some way we can construct this so it queries less data and also won't become outdated? (eg. imagine someone running this in 2026 when they've only just stood up the target clusters)

Author: if we fix the bug for label values, then we can just not specify `start` and `end`. Let's keep this open to see how complicated it will be to fix. If we can fix it in time, i'll just drop the start and end from here

Author: ok, i just added a command which generates always queries the last 1 year; see 87bebecf0e1dc2acaa1fd5d4a96d60e8a2ad7989
